### DISCUSSION TITLE
Does FastAPI support subdomains like Flask?

### DISCUSSION THREAD

**Discussion by Top-Dog** [fastapi/fastapi#8372] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from flask import Flask

app = Flask(__name__)

# Add these definitions to the hosts file:
# local.dev        127.0.0.1
# api.local.dev    127.0.0.1
app.config["SERVER_NAME"] = "local.dev:5000"

@app.route("/")
def home():
    return "Home page"

@app.route("/default-subdomain")
def default_subdomain():
    return "Default subdomain"

@app.route("/", subdomain="api")
def api_home():
    return "API home"
```

### Description

Flask has subdomains for its routes: https://flask.palletsprojects.com/en/2.2.x/api/#url-route-registrations - does FastAPI have something similar? If not, and it has to be done with a Nginx reverse proxy and two separate FastAPI instances (e.g. default subdomain on port 8000 and the api subdomain on port 8001), how would I go about sharing internal state e.g. auth tokens and user id between two different FastAPI apps?

### Operating System

Linux, Windows

### FastAPI Version

0.79.0

### Python Version

3.10.5

**Comment by JarroVGIT** [comment]
FastAPI has no support for subdomains. You don't necessarily have to use two FastAPI instances if you use NGINX. You could just point `subdomain.example.com` to `api.example.com/router/` and `example.com` to `api.example.com`. You just have to make sure that all your subdomain paths are properly mapped, so I would suggest you would use a separate `APIRouter` with a fixed `prefix=`.

It might get confusing, and also note that when people go to `example.com/router/endpoint` that it would be a valid request. This solution is just some sugarcoating in that respect.

**Comment by nicolaipre** [comment]
Any plans on adding this in considering other frameworks (i.e Django, Flask, Sanic, Masonite, Falcon, ASP.Net, Laravel, etc...) already has this implemented?

Edit: Also related to [#445](https://github.com/tiangolo/fastapi/issues/445)

**Comment by arianxplore** [comment]
Yes, Flask implement it already.

**Comment by nicolaipre** [comment]
> 

Hi. I am aware it exists in Flask, but I am asking when it will be implemented in FastAPI.

**Comment by nicolaipre** [comment]
> > > 
> > 
> > 
> > Hi. I am aware it exists in Flask, but I am asking when it will be implemented in FastAPI.
> 
> Understand that it will never be implemented. This is not a responsibility that any web framework should have. There are already reverse proxies for this, such as traefik or nginx

Understand that when doing it this way (e.g. using URI paths for different subdomains in nginx/traefik config), all routes from the same web app will become available on both the main domain and any other subdomains you have set up this way. This is the whole reason why almost all other frameworks have support for this.

I don't understand why you are talking down this feature when all the major and widely used frameworks have this implemented. It has existed for years.

**Comment by nicolaipre** [comment]
Just a quick update on this one - I have recently been looking at Starlette which I understood FastAPI has been built around, and noticed subdomain routing is supported in Starlette as well, with the only difference that you have to specify the full domain and not only the subdomain which would be ideal: https://www.starlette.io/routing/#host-based-routing

This functionality is still on my wishlist for FastAPI.

**Reply by philipbjorge** [reply]
Thanks for sharing @nicolaipre 

This is useful functionality that can help an application avoid additional infrastructure dependencies.

This is useful when hosting on Heroku for example. This can make testing stories easier as well.

Strange to have maintainers be so adamant about something never being supported 🤷‍♀️ — especially when you point out that the underlying, lightweight framework it’s built on supports it 😂 

It’s important to keep an open mind and look towards battle tested frameworks as examples. The project will be stronger for it 🤷‍♀️.

**Reply by nicolaipre** [reply]
Hi.

Thank you for the feedback. Nice to hear it would be useful for you too. 

I decided to make use of [Starlette](https://github.com/encode/starlette) instead since the functionality is supported there, and I personally find the coding style more clean and with optional ways of registering routes, middleware, etc. without being forced to use decorators.

This gives me more flexibility and is a perfect fit for my projects and preferred workflow.

Hopefully FastAPI will add this functionality too at some point. I am however too happy with the way it turned out with Starlette to change back, so I believe I will stick with that.

**Comment by YuriiMotov** [comment, marked as answer]
> **tiangolo** on Feb 10, 2020
> Thanks for the help here!
> 
> ...I would still suggest handling hosts outside of FastAPI itself. Also, it probably should be handled by the same thing that handles HTTPS certificates. So, Traefik, Nginx, or similar.
> 

https://github.com/fastapi/fastapi/issues/445#issuecomment-584180351
